### PR TITLE
keras Model.fit pointnet v1

### PR DESCRIPTION
**Pull request by jackd**
* add `tf.keras.Model.fit` scipt (equivalent to `train`, for demo purposes)
* specifies `axis` of a `squeeze` in model that was only running successfully because of a bug in preprocessing

Test won't pass until #310 is accepted.

There's a lot of shared code between this and `fit` which could be hoisted into a shared file. Will wait for other PRs to get accepted and avoid the merge conflict nightmare...

### CODE REVIEW

**Comment by ataiya** [comment]
Can you clarify your second point? (the bug)

**Comment by jackd** [comment]
[Truncating points](https://github.com/tensorflow/graphics/blob/master/tensorflow_graphics/projects/pointnet/train.py#L73) is along the batch dimension, not the points dimension. For some reason this changes the shape from `(?, 2048, 3)` to `(32, 2034, 3)` (which in the past would have give issues on the final step of training if there isn't a complete batch - though haven't hit that bug recently).

This results in a fully statically known shape prior to squeeze, so squeeze knows which axes it can and cannot remove. Without the bug above, the shape is `(?, 1, F)` - the squeeze doesn't know if it has to apply to the first dimension, tso the output shape is unknown (even the rank).

This and a few other small things address in https://github.com/tensorflow/graphics/pull/311

File: tensorflow_graphics/nn/layer/pointnet.py [hunk 1]
```diff
@@ -208,8 +208,9 @@ def __init__(self, num_classes=40, momentum=.5, dropout_rate=.3):
     """
     super(PointNetVanillaClassifier, self).__init__()
```
Comment (ataiya) [inline tensorflow_graphics/nn/layer/pointnet.py:209]: Feel free to rename this to VanillaClassifier, the fact that it is pointnet is inherited by the module name.
(as you are already changing stuff ;))

File: tensorflow_graphics/projects/pointnet/fit.py [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""keras Model.fit loop for PointNet v1 on modelnet40."""
+# pylint: disable=missing-function-docstring
+import functools
+import tensorflow as tf
+
+from tensorflow_graphics.datasets.modelnet40 import ModelNet40
+from tensorflow_graphics.nn.layer.pointnet import PointNetVanillaClassifier as PointNet
+from tensorflow_graphics.projects.pointnet import augment as augment_lib
```
Comment (ataiya) [inline tensorflow_graphics/projects/pointnet/fit.py:21]: same as in the other review

File: tensorflow_graphics/projects/pointnet/fit_test.py [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Tests of pointnet module."""
+
+import sys
+import tempfile
+
+import tensorflow_datasets as tfds
+from tensorflow_graphics.util import test_case
+from tensorflow_graphics.datasets import testing  # will fail until #310 merge
```
Comment (ataiya) [inline tensorflow_graphics/projects/pointnet/fit_test.py:21]: Sounds good, resolve comment when done.

File: tensorflow_graphics/projects/pointnet/fit_test.py [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+            "--ev_every",
+            "1",
+            "--tb_every",
+            "1",
+            "--logdir",
+            logdir,
+            "--batch_size",
+            str(batch_size),
+        ]
+        from tensorflow_graphics.projects.pointnet import fit  # pylint:disable=import-outside-toplevel
```
Comment (ataiya) [inline tensorflow_graphics/projects/pointnet/fit_test.py:45]: update to same usage from other PR?

File: tensorflow_graphics/projects/pointnet/fit.py [hunk 5]
```diff
@@ -0,0 +1,98 @@
+# Copyright 2020 Google LLC
```
Comment (ataiya) [inline tensorflow_graphics/projects/pointnet/fit.py:1]: Not quite sure you need a google copyright here :)
Try to look at what other TF core contributions from third parties used.

**Comment by ataiya** [comment]
ah yes, of course!
Missed this one as testing with multiple point sizes was not possible.
Now with the latest edits to the testing pipeline this would have been
caught.

File: tensorflow_graphics/projects/pointnet/fit_test.py (hunk 4 above)
Comment (jackd) [reply tensorflow_graphics/projects/pointnet/fit_test.py:45]: Woops, missed this. With the `main` function this import can be hoisted to the top. I feel that's cleaner, and theoretically allows the test to be re-run with different command line args. I think consistency is more important than perfection though. Feelings about converting `train` to a similar scheme? If I pull out `get_datasets` and `get_optimizer` to a shared common lib, there won't be all that much left to through under a `main` function.

Comment (ataiya) [reply tensorflow_graphics/projects/pointnet/fit_test.py:45]: The reason train.py is written this way is that eventually this will become an auto-translate to a jupyter/colab notebook.

So train.py is indeed a script, not an application, and that is why the sys.args has been designed to run in this particular way.

There was a logic ;)

File: tensorflow_graphics/projects/pointnet/fit.py (hunk 5 above)
Comment (jackd) [reply tensorflow_graphics/projects/pointnet/fit.py]: My bad. I've just removed it... I'm not fussed personally. Maybe something to add to https://github.com/tensorflow/graphics/issues/302 ?

File: tensorflow_graphics/projects/pointnet/fit_test.py (hunk 4 above)
Comment (jackd) [reply tensorflow_graphics/projects/pointnet/fit_test.py]: Oooh, I'm guessing the comment blocks are significant here too? They seemd a bit excessive to me, just figured it was a personal taste thing.

I'll convert `fit.py` to the same. `importlib.reload` should be able to allow for testing different configurations.

**Comment by jackd** [comment]
If I'm renaming layers anyway, any chance I can address https://github.com/tensorflow/graphics/issues/314 e.g. `PointNetConv2Layer` -> `PointnetConv2D`

**Comment by ataiya** [comment]
Sure

**Comment by jackd** [comment]
@ataiya this is a lot bigger than I originally planned it to be, but I've just refactored shared functionality between fit/train into `helpers.py`. This now supercedes https://github.com/tensorflow/graphics/pull/311. I'm happy to close 311, but I imagine it might get through the review process faster if you wanted to eliminate that bug ASAP.

Tests waiting https://github.com/tensorflow/graphics/pull/310.

**Comment by ataiya** [comment]
I'll have time to look at things tomorrow, finishing reviews today.
Note helpers should NOT contain any project logic.
It's basically project agnostic functionality that is useful whenever you
have a custom training loop.

**Comment by julienvalentin** [comment]
Hi @jackd !

It seems that you made a lot of progress on this one; what is the current status?

Best.

**Comment by jackd** [comment]
@julienvalentin once #310 is through I'll ensure tests are passing, then should be good for review.

@ataiya should I make a separate `utils.py` file for shared dataset/learning rate functionality shared between `fit` and `train`? They're currently in `helpers.py`, but they're a little project specific...

**Comment by julienvalentin** [comment]
#310 is in! :)

File: tensorflow_graphics/__init__.py [hunk 6]
```diff
@@ -49,4 +49,4 @@
   __all__.remove("version")
 # pylint: enable=g-statement-before-imports,g-import-not-at-top
 
-__version__ = "HEAD"
```
Comment (ataiya) [inline tensorflow_graphics/__init__.py]: The version is auto-generated when setup.py is run. The reason for which it is "HEAD" is because in the source repo it's just the last commit that determines the version.

Comment (jackd) [reply tensorflow_graphics/__init__.py]: I was wondering where I accidentally managed to change that...

File: tensorflow_graphics/projects/pointnet/fit.py [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+optimizer = tf.keras.optimizers.Adam(
+    learning_rate=helpers.decayed_learning_rate(FLAGS.learning_rate) if FLAGS.
+    lr_decay else FLAGS.learning_rate)
+
+model.compile(
+    optimizer=optimizer,
+    loss=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True),
+    metrics=[tf.keras.metrics.SparseCategoricalAccuracy()])
+
+ds_train, ds_test = helpers.get_modelnet40_datasets(FLAGS.num_points,
```
Comment (ataiya) [inline tensorflow_graphics/projects/pointnet/fit.py:55]: this cannot be in helpers, helpers is "project agnostic"

File: tensorflow_graphics/projects/pointnet/helpers.py [hunk 8]
```diff
@@ -16,13 +16,17 @@
...
 import argparse
+import functools
 import os
 import tempfile
 import time
 
 import tensorflow as tf
 import termcolor
 
+from tensorflow_graphics.datasets.modelnet40 import ModelNet40
```
Comment (ataiya) [inline tensorflow_graphics/projects/pointnet/helpers.py:27]: remove

File: tensorflow_graphics/projects/pointnet/helpers.py [hunk 9]
```diff
@@ -154,3 +161,106 @@ def handle_keyboard_interrupt(flags):
   else:
     print("Delete these summaries with: ")
     termcolor.cprint("  rm -rf {}".format(flags.logdir), "red")
+
+
+def preprocess(points,
```
Comment (ataiya) [inline tensorflow_graphics/projects/pointnet/helpers.py:166]: move elsewhere? (file to remain project agnostic)

File: tensorflow_graphics/projects/pointnet/helpers.py [hunk 10]
```diff
@@ -154,3 +161,106 @@ def handle_keyboard_interrupt(flags):
...
+  """
+  points = points[:num_points]
+  if augment_rotate:
+    points = augment.rotate(points)
+  if augment_jitter:
+    points = augment.jitter(points)
+  return points, labels
+
+
+def get_modelnet40_datasets(num_points=2048,
```
Comment (ataiya) [inline tensorflow_graphics/projects/pointnet/helpers.py:204]: move elsewhere? (file to remain project agnostic)

Comment (ataiya) [reply tensorflow_graphics/projects/pointnet/helpers.py:204]: In fact, there's nothing wrong with repeating the logic twice.
The "custom training loop" (my version) vs. keras (your version) are supposed to be self-contained examples. 

Being examples, sharing logic outside the file makes them less clean :)

File: tensorflow_graphics/projects/pointnet/helpers.py [hunk 11]
```diff
@@ -154,3 +161,106 @@ def handle_keyboard_interrupt(flags):
...
+                        augment_rotate=augment_rotate),
+      tf.data.experimental.AUTOTUNE)
+  ds_train = ds_train.batch(batch_size).prefetch(tf.data.experimental.AUTOTUNE)
+  ds_test = ds_test.map(functools.partial(preprocess, num_points=num_points),
+                        tf.data.experimental.AUTOTUNE)
+  ds_test = ds_test.batch(batch_size).prefetch(tf.data.experimental.AUTOTUNE)
+  return ds_train, ds_test
+
+
+def decayed_learning_rate(
```
Comment (ataiya) [inline tensorflow_graphics/projects/pointnet/helpers.py:246]: You could move the parsing of decay flags in here?
(including whether decay is enabled)

File: tensorflow_graphics/projects/pointnet/train.py [hunk 12]
```diff
@@ -128,12 +108,11 @@ def evaluate():
 # ------------------------------------------------------------------------------
 # ------------------------------------------------------------------------------
 
-ds_train, info = ModelNet40.load(split="train", with_info=True)
-num_examples = info.splits["train"].num_examples
-ds_train = ds_train.shuffle(num_examples, reshuffle_each_iteration=True)
-ds_train = ds_train.repeat(FLAGS.num_epochs)
-ds_train = ds_train.batch(FLAGS.batch_size)
-ds_test = ModelNet40.load(split="test").batch(FLAGS.batch_size)
+ds_train, ds_test = helpers.get_modelnet40_datasets(FLAGS.num_points,
```
Comment (ataiya) [inline tensorflow_graphics/projects/pointnet/train.py:111]: not a big fan of this, would like to keep the logic in this file.

File: tensorflow_graphics/projects/pointnet/helpers.py [hunk 13]
```diff
@@ -154,3 +161,106 @@ def handle_keyboard_interrupt(flags):
...
+
+  Returns:
+    (train dataset, test dataset), with batched `(points, labels)` elements.
+  """
+  (ds_train, ds_test), info = ModelNet40.load(as_supervised=True,
+                                              split=("train", "test"),
+                                              with_info=True)
+  num_examples = info.splits["train"].num_examples
+  ds_train = ds_train.shuffle(num_examples, reshuffle_each_iteration=True)
+  if num_epochs != 1:
```
Comment (ataiya) [inline tensorflow_graphics/projects/pointnet/helpers.py:231]: is this necessary? Isn't this a no-op for num_epochs=1?

File: tensorflow_graphics/projects/pointnet/train.py [hunk 14]
```diff
@@ -114,8 +95,7 @@ def evaluate():
   if step % FLAGS.ev_every != 0:
     return evaluate.best_accuracy
   aggregator = tf.keras.metrics.SparseCategoricalAccuracy()
-  for example in ds_test:
-    points, labels = example["points"], example["label"]
+  for points, labels in ds_test:
```
Comment (ataiya) [inline tensorflow_graphics/projects/pointnet/train.py:98]: should be label, as it's batch of points, and batch of labeL 
(this was my mistake)

**Comment by jackd** [comment]
@ataiya I'm happy to duplicate learning rate / preprocessing across examples. The only reason I'm tempted to keep at least data preprocessing separate is for reusability, e.g. it's nice to be able to say with confidence "we used exactly the same data preprocessing as pointnet" without having to copy/paste from examples. I agree `helpers.py` wouldn't be the appropriate place for that, but perhaps a small `preprocessing.py` file?

**Comment by ataiya** [comment]
So load("modelnet/pointnet", ...) would give you the variant with  "exactly
the same data preprocessing as pointnet".
But perhaps, for now, a separate file is sufficient, something like
modelnet_dataset.py or modelnet.py.

For longer term, shouldn't it be located somewhere in datasets/shapenet?
For shapenet, we are using the "configs" to do this, e.g. to generate a
point set version of the shapenet meshes (used by OccNet, DeepSDF, ...)
After all dataset+augmentations can be considered as as (new) specific
sub-version of a dataset.

What do you think?

**Comment by jackd** [comment]
I like to have a clear separation between data stored on disk (from `tfds`) and preprocessing with `tf.data`, e.g. I don't think it's appropriate for jittering to be a part of the builder/config itself, but perhaps a `preprocessing` file alongside it. I guess it depends on whether or not you feel preprocessing is a property of the data or part of the model. I tend to feel its part of the model (though even then it's not clear cut. e.g. I feel the number of points is a property of the the data, while the jittering is a property of the model).

**Comment by ataiya** [comment]
>
> I like to have a clear separation between data stored on disk (from tfds)
> and preprocessing with tf.data. e.g. I don't think it's appropriate for
> jittering to be a part of the builder/config itself, but perhaps a
> preprocessing file alongside it.
>
Yes, but TFDS has a way (or will soon) to store post-processed data *on
disk*.
(or have a cached version already stored, and download that)

> I guess it depends on whether or not you feel preprocessing is a property
> of the data or part of the model. I tend to feel its part of the model
> (though even then it's not clear cut. e.g. I feel the number of points is a
> property of the the data, while the jittering is a property of the model).
>
Ok, go ahead and integrate the changes from the other PR for me to do
another round of comments.
Might be a bit slot due to the NeurIPS deadline.

**Comment by jackd** [comment]
Sorry for the delay on this. I've removed learning rate / data pipelining logic from `helpers` and duplicated it across the two examples (`fit.py` and `train.py`).

There may be a slight performance boost by adding a `Dataset.repeat` at the beginning of the pipeline, and specifying `steps_per_epoch` in fit. This will result in slight bleeding of examples across epochs, but remove the need to fill a new shuffle-buffer from scratch each time. Reducing the buffer size would also reduce the scale of this problem.

The main reason this isn't currently implemented is `tfds.testing.mock_data` doesn't change `info.splits["train"].num_examples`, so `steps_per_epoch` can't be computed consistently across the real run and test run. This -may- be able to be resolved with [upgraded `tfds` version](https://github.com/tensorflow/graphics/issues/470) - it uses `tf.data.experimental.set_cardinality`, so we can compute `num_examples` with `tf.data.experimental.cardinality`, but I'm not sure if the way in which it is set is robust to `mock_data` contexts.

I believe the current build failure is down to #469
